transports/janus_websockets.c

```diff
@@ -1198,56 +1201,63 @@ static int janus_websockets_common_callback(
 						ws_client->buflen = buflen;
 						ws_client->buffer = g_realloc(ws_client->buffer, buflen);
 					}
 					memcpy(ws_client->buffer + LWS_PRE, response, strlen(response));
-					JANUS_LOG(LOG_HUGE, "[%s-%p] Sending WebSocket message (%zu bytes)...\n", log_prefix, wsi, strlen(response));
-					int sent = lws_write(wsi, ws_client->buffer + LWS_PRE, strlen(response), LWS_WRITE_TEXT);
-					JANUS_LOG(LOG_HUGE, "[%s-%p]   -- Sent %d/%zu bytes\n", log_prefix, wsi, sent, strlen(response));
-					if(sent > -1 && sent < (int)strlen(response)) {
+					/* Initialize pending bytes count and buffer offset */
+					ws_client->bufpending = strlen(response);
+					ws_client->bufoffset = LWS_PRE;
+					/* We can get rid of the message */
+					free(response);
+				}
+
+				/* Evaluate amount of data to send according to MESSAGE_CHUNK_SIZE */
+				int amount = ws_client->bufpending <= MESSAGE_CHUNK_SIZE ? ws_client->bufpending : MESSAGE_CHUNK_SIZE;
+				/* Set fragment flags */
+				int flags = lws_write_ws_flags(LWS_WRITE_TEXT, ws_client->bufoffset <= LWS_PRE, ws_client->bufpending <= (size_t)amount);
+				/* Send the fragment with proper flags */
+				int sent = lws_write(wsi, ws_client->buffer + ws_client->bufoffset, (size_t)amount, flags);
+				JANUS_LOG(LOG_HUGE, "[%s-%p]   -- First=%d, Last=%d, Requested=%d bytes, Sent=%d bytes, Missing=%zu bytes\n", log_prefix, wsi, ws_client->bufoffset <= LWS_PRE, ws_client->bufpending <= (size_t)amount, amount, sent, ws_client->bufpending - amount);
+				if(sent < amount) {
+					/* Error on sending, abort operation */
+					JANUS_LOG(LOG_ERR, "Websocket sent only %d bytes (expected %d)\n", sent, amount);
+					ws_client->bufpending = 0;
+					ws_client->bufoffset = 0;
+				}
+				else {
+					/* Fragment successfully sent, update status */
+					ws_client->bufpending -= amount;
+					ws_client->bufoffset += amount;
+					if(ws_client->bufpending > 0) {
+						/* We still couldn't send everything that was left, we'll try and complete this in the next round */
 						/* We couldn't send everything in a single write, we'll complete this in the next round */
-						ws_client->bufpending = strlen(response) - sent;
-						ws_client->bufoffset = LWS_PRE + sent;
-						JANUS_LOG(LOG_HUGE, "[%s-%p]   -- Couldn't write all bytes (%d missing), setting offset %d\n",
+						JANUS_LOG(LOG_HUGE, "[%s-%p]   -- Couldn't write all bytes (%zu missing), setting offset %zu\n",
 							log_prefix, wsi, ws_client->bufpending, ws_client->bufoffset);
 					}
-					/* We can get rid of the message */
-					free(response);
-					/* Done for this round, check the next response/notification later */
-					lws_callback_on_writable(wsi);
-					janus_mutex_unlock(&ws_client->ts->mutex);
-					return 0;
 				}
+				/* Done for this round, check the next response/notification later */
+				lws_callback_on_writable(wsi);
```

Comment:
This seems to be calling `lws_callback_on_writable` even when we successfully sent the last fragment, which shouldn't happen (it will perform one more useless loop).

Reply:
I have preserved the previous behavior (https://github.com/meetecho/janus-gateway/blob/master/transports/janus_websockets.c#L1246).
A quick test showed some issues when sending asyncronous messages when the `writable` call gets moved from that position.

Reply:
Ack!